Draft: Pipewire unified wrappers

columbarius opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
Unify pipewire.c api in the linux-pipewire plugin
<!--- Describe your changes in detail. -->
<!--- If this change includes UI elements, please include screenshots. -->

### Motivation and Context

We want to create a basic pipewire wrapper to share code between multiple obs plugins. This MR should create the basis for discussions to create sth. useful for various video and audio plugins.
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open GitHub Issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
<!--- - New feature (non-breaking change which adds functionality) -->
<!--- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
- Code cleanup (non-breaking change which makes code smaller or more readable)
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [ ] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [ ] My code is not on the master branch.
- [ ] The code has been tested.
- [ ] All commit messages are properly formatted and commits squashed where appropriate.
- [ ] I have included updates to all appropriate documentation.

columbarius committed:
linux-pipewire: Extend format_data

The media API requires it's own format (enum video_format) and
bytes per pixel (bpp).

columbarius committed:
linux-pipewire: Add support for YUY2

columbarius committed:
linux-pipewire: Move stream properties into plugins

This commit makes the stuff in pipewire.c more independent from the
actual plugin by providing a wrapper to create a stream with a set or
properties defined by the plugin.

columbarius commented:
Relevant MRs:
#5377 
#5728
#6207

columbarius commented:
CC: @GeorgesStavracas @dimtpap

kkartaltepe commented on plugins/linux-pipewire/pipewire.h:
```diff
@@ -22,9 +22,16 @@
-void *obs_pipewire_create(int pipewire_fd, int pipewire_node);
+extern const struct pw_stream_events stream_events;
+
+void *obs_pipewire_create(int pipewire_fd, int pipewire_node, const char *name,
```
Why is this void*, the only assignment from this function i see is directly into the correct type.

columbarius replied:
The reason for this decision,  i can think of, is to hide all (for the plugin) "unnecessary" stuff behind a opaque handle.

kkartaltepe replied:
Can we use an opaque type instead of void*, unless there is a real reason we need to use void* (but since this always returns the same type it doesnt seem so?).

columbarius replied:
Why not? But I'd like implement this after we finished designing the whole wrapper api.

columbarius commented:
Moved comment from https://github.com/obsproject/obs-studio/pull/5728#issuecomment-1168993335
> > We separate camera plugin and registry.
> > 
> > * camera-plugin: `init_registry` -> `query_cameras` -> `user selects camera` -> `query_registry for fd and id` -> `create stream like screencast` -> ... -> `unref_registry`
> > * camera-registry: global refcounted struct, which deals with the camera/device-portal. Is created on first `init_registry` call and destroyed if the refcount is 0. It queries the portal for a filedescriptor and opens the connection and tracks all announced cameras.
> 
> In principle, I think separating different moving parts into their own unit is good. My idea with #6645 is that all the code under `pipewire.c` is generic and can be used by all source types - even audio sources.
> 
> I don't know if it would be feasible or even possible, but I think you can already make this `camera-registry` thing completely generic, so it can be reused by the audio source at #6207. It'll probably need to be implemented in `pipewire.c` directly, perhaps as follows:
> 
>     * Split the code that actually initializes PipeWire from `obs_pipewire_create` into a new `obs_pipewire_start`
> 
>     * Add a new vfunc struct `obs_pipewire_device_callbacks` with the regular `device_added`, `device_removed`, `device_changed`, etc callback (please feel free to suggest better naming here)
> 
>     * Add `obs_pipewire_set_device_callbacks()`
> 
> 
> That way the camera portal code will be able to react to PipeWire adding and removing cameras, without having to deal with PipeWire itself.
> 
> What do you think?

I like the idea to build a registry wrapper with these callbacks. I'm not sure a `obs_pipewire_set_device_callbacks()` is feasable but an `obs_pipewire_registry_create()` with these callbacks as an argument (see stream_events here) will be.

> > An alternative would be to not lazyload the registry, but start it when obs is launched. This way we can only add the camera plugin, if obs is allowed to access the portal.
> > If we lazy load I would suggest to work asynchronous with callbacks from the registry to the plugin after it's initialized, but that's just some idea.
> 
> I think the best approach here is to do what I've done with the screencasting code, and completely separate camera portal code from PipeWire itself. What the plugin needs to do is to expose the pw-camera sources all the time, and only query the Device Access portal when the user actually adds a pw-camera source. If camera access is denied, we show an error message in the properties dialog.

I think we might talk about different things here: The screencast portal creates for each stream a new pipewire_fd together with the pipewire_id which the client can use to access only this single stream. Iirc the camera portal only grants access to a single pipewire_fd, which can be used to access pipewire on the host with permissions on the connection only to show the client cameras, etc. In this case the client is responsible to set up a registry to get the pipewire_id of each camera. Can't we just use the provided pipewire_fd together with the pipewire_id, obtained from the registry, and just create a new stream (including it's own pw_core, pw_thread_loop, etc.) with those?    

> > Anyways: should I just continue to rebase the current state, or should I continue with my new proposal?
> 
> Considering what I've suggested above, feel free to continue with your new proposal

This way we could create one wrapper each for pipewire_stream, pipewire_registry (with the suggested callbacks) and pipewire_node (@dimtpap could you extract such a wrapper from your audio work?)

columbarius commented:
Btw. creating `pw_core` and `pw_thread_loop` and such things should be done by static functions inside pipewire.c and shared between the wrapper APIs.

columbarius commented:
Added registry stuff, nothing tested and just a sketch.

columbarius commented:
Does anyone know how to do fa935cc73f1205daf6bd907607b0d35e725a15e7 in nice?

dimtpap commented:
> Does anyone know how to do [fa935cc](https://github.com/obsproject/obs-studio/commit/fa935cc73f1205daf6bd907607b0d35e725a15e7) in nice?

Why not make it a regular struct and pass it to `pw_core_add_listener`

GeorgesStavracas commented on plugins/linux-pipewire/pipewire.h:
```diff
@@ -27,8 +27,11 @@
 
 typedef struct _obs_pipewire_data obs_pipewire_data;
 
+extern const struct pw_stream_events stream_events;
+
```
Please don't. **If** allowing plugins to override stream events is a good idea - something I'm not convinced of yet - then it's better to write it as follows:

```c
struct pw_stream_events obs_pipewire_get_default_stream_events (void);
```

then in the C file it returns a copy of the structure:

```c
struct pw_stream_events obs_pipewire_get_default_stream_events (void)
{
	return stream_events;
}
```

Doing this will at least prevent plugins from accidentally screwing up other plugins.

GeorgesStavracas commented on plugins/linux-pipewire/pipewire.c:
```diff
@@ -46,6 +46,26 @@ struct obs_pw_version {
 
+typedef struct _obs_pipewire_data obs_pipewire_data;
+
+// This struct contains the shared fields between all obs_pipewire wrappers
+// The C Standard guarantees that the memory layout off these parts of the
+// struct are identical and as such we can cast any extended struct to this
+// and use the fields without issues. Any change here has to be added to the
+// extended structs and the order of fields has to be preserved.
```
I don't think this is a good approach for sharing struct fields. Traditionally, if you want `obs_pipewire_stream_data` to extend `obs_pipewire_data`, you'd do something like:

```c
struct _obs_pipewire_data {
	... fields ...
};

struct _obs_pipewire_stream_data {
	obs_pipewire_data base;

	... other fields ...
};
```

Then, in the places where you want to treat `obs_pipewire_stream_data`  as a `obs_pipewire_data`, you'd do:

```c
obs_pipewire_data *data = &stream_data->base;
```

GeorgesStavracas reviewed (changes requested):
I think this PR needs some work, but ultimately what it tries to achieve is desired for the maintainability of the plugin. Technical details aside, I think the internal API here is a bit odd in some places. What I had in mind with device registry callbacks and whatnot was something like this:

```c
struct obs_pipewire_device_callbacks {
	void (*device_added)(void *user_data, obs_pipewire_device *device);
	void (*device_removed)(void *user_data, obs_pipewire_device *device);
};

void obs_pipewire_add_device_callbacks(
	obs_pipewire_data *obs_pw, 
	const struct obs_pipewire_device_callbacks *device_callbacks);

void obs_pipewire_remove_device_callbacks(
	obs_pipewire_data *obs_pw, 
	const struct obs_pipewire_device_callbacks *device_callbacks);
```

To properly do this, we would need to split the initialization of `obs_pipewire` in 2 steps:

```c
obs_pipewire_data * obs_pipewire_create(int pipewire_fd);

obs_pipewire_stream *
obs_pipewire_connect_stream(
	obs_pipewire_data *obs_pw,
	uint32_t stream_node,
	const char *stream_name,
	struct pw_properties *stream_properties);
```

The screencast portal code would look more or less like this:

```c
	capture->obs_pw = obs_pipewire_create(pipewire_fd);
	capture->pipewire_stream = obs_pipewire_connect_stream(capture->obs_pw,
		capture->pipewire_node, "OBS Studio",
		pw_properties_new(PW_KEY_MEDIA_TYPE, "Video",
				  PW_KEY_MEDIA_CATEGORY, "Capture",
				  PW_KEY_MEDIA_ROLE, "Screen", NULL)
	obs_pipewire_stream_set_cursor_visible(capture->pipewire_stream,
					capture->cursor_visible);
```


The camera portal code would look more or less like this:

```c

static const struct obs_pipewire_device_callbacks camera_device_callbacks = {
	camera_portal_device_added,
	camera_portal_device_remove,
};

...

	camera->obs_pw = obs_pipewire_create(pipewire_fd);
	obs_pipewire_add_device_callbacks(camera->obs_pw, &camera_device_callbacks);

... (every time someone selects a camera) ...

	capture->pipewire_stream = obs_pipewire_connect_stream(camera->obs_pw,
		device->pipewire_node, "OBS Studio",
		pw_properties_new(PW_KEY_MEDIA_TYPE, ...,
				  PW_KEY_MEDIA_CATEGORY, ...,
				  PW_KEY_MEDIA_ROLE, "Screen", NULL));
```

GeorgesStavracas committed:
linux-pipewire: Return actual type in obs_pipewire_create

Instead of returning an opaque void* pointer, return the leaf
C type obs_pipewire_data*.

This commit introduces no functional changes.

GeorgesStavracas committed:
linux-pipewire: Cleanup D-Bus proxy on unload

When the linux-pipewire plugin is unloaded, make sure to cleanup
the D-Bus proxy for the ScreenCast portal too. This effectively
doesn't change anything, but it's always good to keep up with the
code hygiene.

GeorgesStavracas committed:
linux-pipewire: Inline play_pipewire_stream()

This separation of obs_pipewire_create() and play_pipewire_stream()
was an artifact of how the original code was written, and there is
no reason to keep this separation anymore. Inlining it will help
future commits too.

GeorgesStavracas committed:
linux-pipewire: Remove unnecessary struct field

We don't actually need to restore the node id anymore, since this
is handled by the screencast-portal.c code now.

Remove the pipewire_node field, and just create the stream directly
from the passed now.

GeorgesStavracas committed:
linux-pipewire: Split initialization of core and streams

This commit introduces an important distinction in the initialization
process that is essential to future camera and audio work: it splits
creating and connecting to the PipeWire socket, from connecting to
specific PipeWire nodes.

Right now, for the only consumer in existance - the screencast portal
code - this distinction is irrelevant, but from an API prespective it
just makes sense to model it this way.

GeorgesStavracas committed:
linux-pipewire: Rename obs_pipewire_data to obs_pipewire

Next commits will introduce new obs_pipewire_* types, so this
renaming will make it slightly easier to read the code with
different types.

GeorgesStavracas committed:
linux-pipewire: Trivially shuffle some code around

This will make code easier to follow after next commit.

GeorgesStavracas committed:
linux-pipewire: Separate PipeWire core and streams

Sorry, this is a painful commit to review :(

Until now, the only consumer of the PipeWire code is the screen
cast portal code. This portal code only ever has one PipeWire
connection, and one PipeWire stream, every time a monitor or a
window is selected. This is reflected on obs_pipewire, which only
handles a single stream for any given connection.

For cameras and audio, however, a single PipeWire connection can
and most likely always will provide multiple streams. For example,
computers with multiple webcams will have one PipeWire connection
reporting multiple PipeWire nodes, one node for each camera.

This commit breaks this one-stream-per-connection assumption, and
makes each call to obs_pipewire_connect_stream() return a completely
separate stream object.

A new structure, obs_pipewire_stream, is introduced to represent
a PipeWire stream. The screencast portal code continues to only
ever have one connection and one stream, and that's made explicit
in the code.

GeorgesStavracas committed:
linux-pipewire: Track streams

Make obs_pipewire hold a list of streams created by itself, and
destroy these streams when closing the connection.

GeorgesStavracas committed:
linux-pipewire: Store source when creating streams

We'll need to peek information about the source to determine how
to process frames, so store it for later usage.

GeorgesStavracas committed:
linux-pipewire: Move format info to streams

So far we've been treating format info on a per-connection basis,
but now that a single connection is capable of hosting multiple
streams, and each stream might negotiate a different format, it
is necessary to move format info to each stream individually.

GeorgesStavracas committed:
linux-pipewire: Factor out format data in a struct

And use it instead of exhaustively listing each field of the
anonymous struct into the arguments of the lookup function.

GeorgesStavracas committed:
linux-pipewire: Move portal signal subcription to portal.c

This will be reused by the camera portal code.

GeorgesStavracas committed:
linux-pipewire: Allow passing registry events on creation

This will allow consumers of obs_pipewire to handle the device
registry themselves. This is not useful for the screencast code
since there's always only one node in the registry, but it will
be used by the camera and audio code to list hardware devices.

GeorgesStavracas committed:
linux-pipewire: Add obs_pipewire_roundtrip()

Generalize the async server roundtrip wait so that future consumers
can, for example, wait for all devices to be listed before acting
on the registry.

columbarius committed:
linux-pipewire: Introduce obs_pipewire_stream_data

This struct tracks the related obs object. To share common code with the
virtualcamera plugin we need to be able to deal with obs_source_t and
obs_output_t.

columbarius committed:
linux-pipewire: Implement output_stream

This allows us to write a plugin exporting the buffers via PipeWire

columbarius committed:
linux-pipewire: Add support for async sources

Cameras and audio streams will use the async output routines, but
not screencasting. This needs to be handled differently in the
processing callback.

Co-authored-by: Georges Basile Stavracas Neto <georges.stavracas@gmail.com>

columbarius commented:
Added the case of an output stream to pipewire.c in planing of the virtualcamera pluging.

columbarius commented:
@GeorgesStavracas I'm pretty happy with your ideas! From my side it's ready for review.

columbarius commented:
@GeorgesStavracas I'm not sure about the api design. On one hand I'd like to have a single function for `obs_pipewire_stream_pause` on the other hand audio won't need a `_get_width` function. 

Would you prefere one api for video_in, video_out, audio_in or one unified one, created with a 
```
struct {
    type,
    union {
        video_in,
        ...,
    }
}
```
Thing and internally we do a lot of switch case?

columbarius commented:
I went this weekend on a totally different direction using a struct of function pointers holding the implementation of certain features for each backend which wants to implement the `obs_pipewire_stream` thing/interface. The results are quite nice.
[Here](https://github.com/columbarius/obs-studio/tree/pipewire-unified-dev) is a branch containing rebased versions of #5377 and #5728. 

@GeorgesStavracas Please take a glance at it, if we want to go with this kind of api (`pipewire-internal.h`)

@dimtpap Please look if extending the `impl` struct with audio functions would help you.

wtay commented:
I rebased the branch and added some small fixes: https://github.com/wtay/obs-studio/tree/pipewire-unified-dev

GeorgesStavracas commented:
This pull request has become too chaotic and unmanageable. I'll start breaking it in smaller chunks, and will handle it.

I think something went wrong with my rebase, half of the commits don't even build. I'll fix that too.

columbarius commented:
@GeorgesStavracas please ping me when I should start rebasing this including fixes from @wtay  onto #8153 
Nit: you linked #5614 instead of this MR there.

Same is valid if a review from me is helpful.

PatTheMav commented:
@GeorgesStavracas What's the intended plan for this PR? Is there a roadmap of changes that need to be merged first and does this PR still make sense in its current form (I guess the change itself is desired, just not the way it's implemented in the PR)?
